### ISSUE TITLE
SASRec 模型疑問，Key masking 以及 Query masking

### ISSUE THREAD

**Issue by R7788380** [ZiyaoGeng/RecLearn#19]
作者您好，我在閱讀您的代碼時，有一個地方怎麼想都想不通，想請教您的意見

程式碼如下，當 x 通過權重 wk 和 wk 之後會得到 key 和 query，接下來就會通過下面兩個 masking，我很好奇為何這邊要針對 key 和 query 做 masking。

```
# Key Masking
    key_masks = tf.sign(tf.abs(tf.reduce_sum(k, axis=-1)))  # (None, seq_len)
    key_masks = tf.tile(tf.expand_dims(key_masks, 1), [1, q.shape[1], 1])  # (None, seq_len, seq_len)

    paddings = tf.ones_like(scaled_att_logits) * (-2 ** 32 + 1)
    outputs = tf.where(tf.equal(key_masks, 0), paddings, scaled_att_logits)  # (None, seq_len, seq_len)

# Query Masking
    query_masks = tf.sign(tf.abs(tf.reduce_sum(q, axis=-1)))  # (None, seq_len)
    query_masks = tf.tile(tf.expand_dims(query_masks, -1), [1, 1, q.shape[1]])  # (None, seq_len, seq_len)
    outputs *= query_masks
```

依照第一行 `tf.sign(tf.abs(tf.reduce_sum(k, axis=-1)))` 的作用是將 key embedding 中的最後一個維度進行總和、絕對值以及 `tf.sign`，返回的值為 0 和 1，通常將 embedding 加總後不太可能剛好是 0 ，一定是正負數，通過 `tf.abs` 後只剩下正數，然後通過 `tf.sign` 就會形成全部都是 1 的矩陣，也就是上面的 `key_masks`，請問為何這裏要進行加總判斷的動作 ?

**Comment by ZiyaoGeng** [comment]
我参考的是Transformer的写法，似乎确实没什么作用，您可以考虑考虑，[](https://github.com/Kyubyong/transformer/blob/master/modules.py)
